aws/nat_gateway.go

```diff
@@ -0,0 +1,156 @@
+	if ngw == nil {
+		return false
+	}
+
+	if ngw.CreateTime != nil && excludeAfter.Before(*ngw.CreateTime) {
+		return false
+	}
+
+	return config.ShouldInclude(
+		getNatGatewayName(ngw),
+		configObj.NatGateway.IncludeRule.NamesRegExp,
+		configObj.NatGateway.ExcludeRule.NamesRegExp,
+	)
+}
+
+func getNatGatewayName(ngw *ec2.NatGateway) string {
+	for _, tag := range ngw.Tags {
+		if aws.StringValue(tag.Key) == "Name" {
+			return aws.StringValue(tag.Value)
+		}
+	}
+	return ""
+}
+
+func nukeAllNatGateways(session *session.Session, identifiers []*string) error {
+	region := aws.StringValue(session.Config.Region)
+
+	svc := ec2.New(session)
+
+	if len(identifiers) == 0 {
+		logging.Logger.Infof("No Nat Gateways to nuke in region %s", region)
+		return nil
+	}
+
+	// There is no bulk delete nat gateway API, so we delete the batch of nat gateways concurrently using go routines.
+	logging.Logger.Infof("Deleting Nat Gateways in region %s", region)
+	wg := new(sync.WaitGroup)
+	wg.Add(len(identifiers))
+	errChans := make([]chan error, len(identifiers))
+	for i, ngwID := range identifiers {
+		errChans[i] = make(chan error, 1)
+		go deleteNatGatewayAsync(wg, errChans[i], svc, ngwID)
+	}
+	wg.Wait()
+
+	// Collect all the errors from the async delete calls into a single error struct.
+	var allErrs *multierror.Error
+	for _, errChan := range errChans {
+		if err := <-errChan; err != nil {
+			allErrs = multierror.Append(allErrs, err)
+			logging.Logger.Errorf("[Failed] %s", err)
+		}
+	}
+	finalErr := allErrs.ErrorOrNil()
+	if finalErr != nil {
+		return errors.WithStackTrace(finalErr)
+	}
+
+	// Now wait until the NAT gateways are deleted
+	err := retry.DoWithRetry(
+		logging.Logger,
+		"Waiting for all NAT gateways to be deleted.",
+		// Wait a maximum of 5 minutes: 10 seconds in between, up to 30 times
+		30, 10*time.Second,
+		func() error {
+			areDeleted, err := areAllNatGatewaysDeleted(svc, identifiers)
+			if err != nil {
+				return errors.WithStackTrace(retry.FatalError{Underlying: err})
+			}
+			if areDeleted {
+				return nil
+			}
+			return fmt.Errorf("Not all NAT gateways deleted.")
+		},
+	)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+	for _, ngwID := range identifiers {
+		logging.Logger.Infof("[OK] NAT Gateway %s was deleted in %s", aws.StringValue(ngwID), region)
+	}
+	return nil
+}
+
+// areAllNatGatewaysDeleted returns true if all the requested NAT gateways have been deleted. This is determined by
+// querying for the statuses of all the NAT gateways, and checking if AWS knows about them (if not, the NAT gateway was
+// deleted and rolled off AWS DB) or if the status was updated to deleted.
+func areAllNatGatewaysDeleted(svc *ec2.EC2, identifiers []*string) (bool, error) {
+	// NOTE: we don't need to do pagination here, because the pagination is handled by the caller to this function.
+	resp, err := svc.DescribeNatGateways(&ec2.DescribeNatGatewaysInput{NatGatewayIds: identifiers})
```

Comment:
NIT: is there a max number of NAT gateway IDs that can be passed in?

Reply:
The max is 1000, which I doubt would be hit so I think it's safe to leave this as is.